src/highlights/highlight_group_store.ts

```diff
@@ -0,0 +1,157 @@
+        options.fontStyle = "italic";
+    }
+    if (uiAttrs.bold !== undefined) {
+        options.fontWeight = "bold";
+    }
+    if (uiAttrs.strikethrough !== undefined) {
+        options.textDecoration = "line-through solid";
+    }
+    if (uiAttrs.underline !== undefined) {
+        options.textDecoration = `underline ${specialColor} solid`;
+    }
+    if (uiAttrs.undercurl !== undefined) {
+        options.textDecoration = `underline ${specialColor} wavy`;
+    }
+    return options;
+}
+
+function normalizeThemeColor(color: string | ThemeColor | undefined): string | ThemeColor | undefined {
+    if (typeof color === "string" && color.startsWith("theme.")) {
+        color = new ThemeColor(color.slice(6));
+    }
+    return color;
+}
+
+function normalizeDecorationConfig(config: ThemableDecorationRenderOptions): ThemableDecorationRenderOptions {
+    const newConfig: ThemableDecorationRenderOptions = { ...config };
+    newConfig.backgroundColor = normalizeThemeColor(newConfig.backgroundColor);
+    newConfig.borderColor = normalizeThemeColor(newConfig.borderColor);
+    newConfig.color = normalizeThemeColor(newConfig.color);
+    newConfig.outlineColor = normalizeThemeColor(newConfig.outlineColor);
+    newConfig.overviewRulerColor = normalizeThemeColor(newConfig.overviewRulerColor);
+    return newConfig;
+}
+
+export class HighlightGroupStore implements Disposable {
+    private disposables: Disposable[] = [];
+
+    /**
+     * HL group id to text decorator
+     */
+    private highlighIdToDecorator: Map<number, TextEditorDecorationType> = new Map();
+    /**
+     * Store configuration per decorator
+     */
+    private decoratorConfigurations: Map<TextEditorDecorationType, ThemableDecorationRenderOptions> = new Map();
+
+    // Treat all colors mixed with Visual as Visual to avoid defective rendering due to disjointed decoration ranges.
+    private visualHighlightId?: number;
+    private visualHighlightIds: number[] = [];
+
+    private configuration: HighlightConfiguration;
+
+    constructor() {
+        const highlights: HighlightConfiguration["highlights"] = {};
+        for (const [key, opts] of Object.entries(config.highlights)) {
+            highlights[key] = normalizeDecorationConfig(opts);
+        }
+        this.configuration = { highlights };
+    }
+
+    private createDecoratorForHighlightId(id: number, options: ThemableDecorationRenderOptions): void {
+        if (options.borderColor != null && options.border == null) {
+            options.border = "1px solid";
+        }
+        const decorator = window.createTextEditorDecorationType({
+            ...options,
+            rangeBehavior: DecorationRangeBehavior.ClosedClosed,
+        });
+        this.decoratorConfigurations.set(decorator, options);
+        this.highlighIdToDecorator.set(id, decorator);
+        this.disposables.push(decorator);
+    }
+
+    public addHighlightGroup(id: number, attrs: VimHighlightUIAttributes, groups: string[]): void {
```

Comment:
don't need to repeat the name of the container in its methods. but docstrings are useful.

```suggestion
    /** Adds a highlight group. */
    public add(id: number, attrs: VimHighlightUIAttributes, groups: string[]): void {
```

Reply:
Coincide with my original idea